### PR TITLE
[DO NOT MERGE][Distributed][Refactoring] Improvements to scheduling code

### PR DESCRIPTION
**Pull request by gradientsky**
*Description of changes:*
- Disable forkserver forcing by default
- moved file utils
- Extracted all resource management into managers.py
- Stability improvements

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-820-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-820/1/index.html

**Comment by szha** [comment]
Job PR-820-7 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-820/7/index.html

**Comment by gradientsky** [comment]
Tested using `nthreads_per_trial=1` (all cores are used - worst case scenario with the highest probability of hanging) + `time_limits=60*45`. No hanging observed on multiple runs.

**Comment by szha** [comment]
Job PR-820-8 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-820/8/index.html

File: core/src/autogluon/core/scheduler/fifo.py [hunk 1]
```diff
@@ -363,7 +367,8 @@ def add_job(self, task, **kwargs):
             self.scheduled_tasks.append(task_dict)
 
     def _clean_task_internal(self, task_dict):
-        task_dict['ReporterThread'].join()
+        while task_dict['ReporterThread'].is_alive():
+            task_dict['ReporterThread'].join(timeout=60)
```
Comment (Innixma) [inline core/src/autogluon/core/scheduler/fifo.py:371]: maybe add a comment describing what this does and why its needed

File: core/src/autogluon/core/scheduler/jobs.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+                partial(self._run_dist_job, self.task.task_id),
+                self.task.fn,
+                self.task.args,
+                self.task.resources.gpu_ids
+            )
+            # Fast job submission from multiple threads causes lost jobs and hangs reporter thread.
+            # Keep the values above 200ms combined
+            time.sleep(0.15)
+            job.add_done_callback(_release_resource_callback)
+            time.sleep(0.15 + random.random())
```
Comment (Innixma) [inline core/src/autogluon/core/scheduler/jobs.py:54]: random.random() is pretty large (0.5s on average), perhaps we make this smaller? Fine with keeping it as is if there is risk when reducing though.

File: core/src/autogluon/core/scheduler/jobs.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+        return job
+
+    @staticmethod
+    def _run_dist_job(task_id, fn, args, gpu_ids):
```
Comment (Innixma) [inline core/src/autogluon/core/scheduler/jobs.py:59]: would it make sense to be a `@classmethod` so you don't need to specify DistributedJobRunner in your inner calls (`._worker`, `.__print`)?

File: core/src/autogluon/core/scheduler/managers.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import multiprocessing as mp
+
+from autogluon.core.scheduler.remote import RemoteManager
+from autogluon.core.scheduler.resource import DistributedResourceManager
+
+
+class TaskManagers(object):
+    LOCK = mp.RLock()
```
Comment (Innixma) [inline core/src/autogluon/core/scheduler/managers.py:8]: Difference between RLock and Lock?

File: core/src/autogluon/core/scheduler/remote/remote_manager.py [hunk 5]
```diff
@@ -40,66 +42,72 @@ def __new__(cls):
...
-        ip_addrs = [ip_addrs] if isinstance(ip_addrs, str) else ip_addrs
-        remotes = []
-        for node_ip in ip_addrs:
-            if node_ip in cls.NODES.keys():
-                logger.warning('Already added remote {}'.format(node_ip))
-                continue
-            port = cls.get_port_id()
-            remote = Remote(node_ip, port)
-            with cls.LOCK:
+        with write_lock(cls.LOCK):
```
Comment (Innixma) [inline core/src/autogluon/core/scheduler/remote/remote_manager.py:79]: Why are you wrapping read lock into the `write_lock` function? (I am unfamiliar with these functions)

File: core/tests/unittests/autogluon/core/utils/test_multiprocessing_utils.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        self.__start_and_join_threads(threads)
+
+        ## The two last writers should go first -- after the first reader and
+        ## before the second and the third reader
+
+        self.assertEqual([1], threads[1].buffer_read)
+        self.assertEqual([1, 2, 3], threads[3].buffer_read)
+        self.assertEqual([1, 2, 3], threads[4].buffer_read)
+        self.assert_(threads[0].exit_time < threads[1].entry_time)
+        self.assert_(threads[1].exit_time <= threads[2].entry_time)
```
Comment (Innixma) [inline core/tests/unittests/autogluon/core/utils/test_multiprocessing_utils.py:200]: nit: doc build deprecation warning spam:

```
tests/unittests/autogluon/core/utils/test_multiprocessing_utils.py::RWLockTestCase::test_many_writers_priority

  /var/lib/jenkins/workspace/workspace/autogluon-core-py3/core/tests/unittests/autogluon/core/utils/test_multiprocessing_utils.py:200: DeprecationWarning: Please use assertTrue instead.

    self.assert_(threads[1].exit_time <= threads[2].entry_time)
```

**Review by Innixma — COMMENTED** [review]
Looks good, but I don't have in-depth knowledge of the schedulers to determine if there are defects in the changes.

File: core/src/autogluon/core/scheduler/managers.py (hunk 4 above)
Comment (gradientsky) [reply core/src/autogluon/core/scheduler/managers.py:8]: RLock == reentrant lock same thread can re-enter the lock if needed.

File: core/src/autogluon/core/scheduler/remote/remote_manager.py (hunk 5 above)
Comment (gradientsky) [reply core/src/autogluon/core/scheduler/remote/remote_manager.py:79]: These are contexts. What it does: it acquires write lock, runs the function inside, then releases the lock. This is custom `RWLock`, which non-blocking for read operations, but will block read/write if write lock is acquired.

File: core/src/autogluon/core/scheduler/fifo.py (hunk 1 above)
Comment (gradientsky) [reply core/src/autogluon/core/scheduler/fifo.py:371]: I don't think this is needed anymore - these are left-overs from earlier work. I'll revert it back

File: core/src/autogluon/core/scheduler/jobs.py (hunk 2 above)
Comment (gradientsky) [reply core/src/autogluon/core/scheduler/jobs.py:54]: `0.1` would work, but I would keep it as-is to have some safety margin.

File: core/src/autogluon/core/scheduler/jobs.py (hunk 3 above)
Comment (gradientsky) [reply core/src/autogluon/core/scheduler/jobs.py:59]: updated

File: core/tests/unittests/autogluon/core/utils/test_multiprocessing_utils.py (hunk 6 above)
Comment (gradientsky) [reply core/tests/unittests/autogluon/core/utils/test_multiprocessing_utils.py:200]: good catch, fixed

**Review by Innixma — APPROVED** [review]

**Comment by szha** [comment]
Job PR-820-20 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-820/20/index.html

**Comment by gradientsky** [comment]
Non-deterministically hangs in text application - closing for rework

**Pull request closed without merging**